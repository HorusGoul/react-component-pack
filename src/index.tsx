import * as React from 'react';

export interface WithChildrenProps {
  children: React.ReactNode;
}

export type PackComponentType<T extends WithChildrenProps = WithChildrenProps> = React.ElementType<T>;

export function pack(children: React.ReactNode = null, ...components: PackComponentType[]) {
  if (!components.length) {
    return children as JSX.Element;
  }

  const [Component, ...rest] = components;

  return <Component>{pack(children, ...rest)}</Component>;
}

export function createPack(...components: PackComponentType[]) {
  return function PackComponent({ children }: WithChildrenProps) {
    return pack(children, ...components);
  };
}

