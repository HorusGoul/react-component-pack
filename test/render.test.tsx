import * as React from 'react';
import { createPack } from '../src';
import { WithChildrenComponent } from './utils/test-components';
import * as renderer from 'react-test-renderer';

describe('it', () => {
  it('renders without crashing a pack of one component', () => {
    const Pack = createPack(WithChildrenComponent);

    const tree = renderer.create(<Pack>Test</Pack>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing a pack of three components', () => {
    const Pack = createPack(
      WithChildrenComponent,
      WithChildrenComponent,
      WithChildrenComponent
    );

    const tree = renderer.create(<Pack>Test</Pack>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
