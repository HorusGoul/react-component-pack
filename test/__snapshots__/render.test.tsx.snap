// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders without crashing a pack of one component 1`] = `
<div>
  Test
</div>
`;

exports[`it renders without crashing a pack of three components 1`] = `
<div>
  <div>
    <div>
      Test
    </div>
  </div>
</div>
`;
